### PR TITLE
add picnn + convexity unit test

### PR DESCRIPTION
**Pull request by y-prudent**

### CODE REVIEW

File: cnqr/_src/convex.py [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+  init_std: float = 1e-1
+  init_fn: Callable = jax.nn.initializers.normal
+  act_fn_x: Callable = nn.softplus  # alternative -> relu
+  act_fn_z: Callable = nn.softplus  # alternative -> relu
+  dim_data: int = 2
+  dim_context: int = 1
+  context_dense_layer: str = nn.Dense  # TODO: try with BjorckDense
+  gaussian_map: Tuple[jnp.ndarray, jnp.ndarray] = None
+
+  def setup(self):
```
Comment (Algue-Rythme) [inline cnqr/_src/convex.py]: TODO: remove `setup` and put eveything in `__call__` with `nn.compact`.

File: cnqr/_src/convex.py [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+
+    # ===== Context network =====
+    # (now simple dense, then will be replaced with lipschitz)
+    f_x = []  # dense layers (with bias), init: weights -> identity, bias -> 0
+    for i in range(0, self.num_hidden + 1):
+      output_dim = 1 if i == self.num_hidden else self.dim_hidden[i]
+      f_x.append(
+          self.context_dense_layer(
+              output_dim,
+              name=f'f_x_dense_{i}',
```
Comment (Algue-Rythme) [inline cnqr/_src/convex.py]: Good idea to enforce custom names.

File: cnqr/_src/convex.py [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+              kernel_init=self.init_fn(self.init_std),
+              bias_init=nn.initializers.constant(0.0),
+              use_bias=True,
+          )
+      )
+    self.f_v = f_v
+
+    f_w = []  # dense layers (with bias), init: weights -> approx(0), bias -> 0
+    for i in range(0, self.num_hidden + 1):
+      output_dim = 1 if i == self.num_hidden else self.dim_hidden[i]
```
Comment (Algue-Rythme) [inline cnqr/_src/convex.py]: There is code duplication for `f_u` and `f_w`, and even for `f_v` in some extent.  
I suggest you create lists `output_dims_uw = self.dim_hidden + [1]` and `output_dims_v = self.dim_hidden + [self.dim_data]`.
Then you can replace `for i in range(0, self.num_hidden + 1):` by `for i, output_dim in enumerate(output_dims_uw):` (for example).

File: cnqr/_src/convex.py [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+    z = self.q(y) * jnp.ones(shape=x.shape)
+    for i in range(self.num_hidden + 2):
+      # context network
+      x = self.act_fn_x(self.f_x[i](x))
+      # context modulators
+      u = self.f_u[i](x)
+      v = self.f_v[i](x)
+      w = self.f_w[i](x)
+      # convex network
+      z = self.act_fn_z(self.w_U[i](z * nn.softplus(u)) + self.w_V[i](y * v) + w)
```
Comment (Algue-Rythme) [inline cnqr/_src/convex.py]: Add the variable `h_k` to split computation of `self.w_U[i](z * nn.softplus(u)) + self.w_V[i](y * v) + w` and  `self.act_fn_z`.

File: cnqr/_src/convex.py [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+  def __call__(self, x, y):
+    """
+    Args:
+        x: Context
+        y: Observation from source distribution
+    """
+    # TODO: put everything in the call function (delete setup)
+    x = self.phi(x)
+    z = self.q(y) * jnp.ones(shape=x.shape)
+    for i in range(self.num_hidden + 2):
```
Comment (Algue-Rythme) [inline cnqr/_src/convex.py]: Since `i` is never explicitly used (it is only used to access elements in list) I suggest to replace this line with somehting along :

```python
for layers in zip(self.f_x, self.f_u, self.f_v, self.f_w, self.w_U, self.w_V):
   f_x, f_u, f_v, f_w, w_U, w_V = layers
```

The other advantage is that you don't need to rely on ad-hoc computations like `self.num_hidden + 2` to ensure you are doing the right number of steps.  

If my understanding is correct you create `self.num_hidden + 1` layers in the setup method, but here you iterate over `self.num_hidden + 2` iterations ? Can you explain this ?

File: cnqr/_src/convex.py [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+      # context network
+      x = self.act_fn_x(self.f_x[i](x))
```
Comment (Algue-Rythme) [inline cnqr/_src/convex.py]: Warning, in the paper [Supervised Training of Conditional Monge Maps](https://arxiv.org/pdf/2206.14262.pdf) they use `x_k`, not `x_{k+1}` so this line should be at the end of the `for loop`.

**Comment by Algue-Rythme** [comment]
Good job !

We need to re-factor with `nn.compact` and add other tests.

File: cnqr/_src/convex.py [hunk 7]
```diff
@@ -84,76 +98,48 @@ def setup(self):
...
-                self.context_dense_layer(
-                    output_dim,
-                    name=f'f_u_dense_{i}',
-                    kernel_init=self.init_fn(self.init_std),
-                    bias_init=nn.initializers.constant(1.0),
-                    use_bias=True,
+        f = dict()  # f_u, f_v, f_w
+        output_dims_uw = list(self.dim_hidden) + [1]
+        output_dims_v = [self.dim_data] * num_hidden + [1]
+        for module_name, output_dims in zip(['f_u', 'f_v', 'f_w'],
```
Comment (Algue-Rythme) [inline cnqr/_src/convex.py]: maybe there is too much factorization now: you are using `nn.initializers.constant(1.0)` for every bias in every function, which in not compliant with the initialization scheme proposed.  
  
I believe keeping 3 for loops is better for clarity (and keeping track of everything), my suggestion was just to use `enumerate` instead of the one-liner `if i == self.num_hidden else`.

File: cnqr/_src/convex.py [hunk 8]
```diff
@@ -178,37 +163,19 @@ def setup(self):
...
+        x = phi(x)
+        z = q(y) * jnp.ones(shape=x.shape)
+        for i in range(num_hidden + 1):
             # context modulators
-            u = self.f_u[i](x)
-            v = self.f_v[i](x)
-            w = self.f_w[i](x)
+            u = f['f_u'][i](x)
+            v = f['f_v'][i](x)
+            w = f['f_w'][i](x)
```
Comment (Algue-Rythme) [inline cnqr/_src/convex.py]: The goal was to get rid of `i` here ;)

File: cnqr/_src/convex.py [hunk 9]
```diff
@@ -79,103 +106,104 @@ def __call__(self, x, y):
...
+            w_Vk = w_V(features=output_dims_V[i], name=f'w_V_dense_{i}')
+            h = w_Uk(z * self.pos_act_fn(u)) + w_Vk(y * v) + w
+            z = self.sigma_act_fn(h)
             # context network
-            if i < num_hidden:
-                x = self.act_fn_x(f_x[i](x))
+            if i < num_steps - 1:
+                x = f_x(features=output_dims_x[i], name=f'f_x_dense_{i}')(x)
+                x = self.tau_act_fn(x)
         return jnp.squeeze(z)
```
Comment (Algue-Rythme) [inline cnqr/_src/convex.py]: A bit dangerous IMO: if the batch size is equal to 1, the corresponding dimension will be removed, which may cause issues since the output shape is no longer `(batch_size,)` as expected. It is better to use a good old' fashion `reshape` here.

File: cnqr/_src/convex.py [hunk 10]
```diff
@@ -16,58 +19,82 @@
...
-
-    @staticmethod
-    def _identity_initializer(key: KeyArray,
-                              shape: core.Shape,
-                              dtype: DTypeLikeInexact = jnp.float_) -> Array:
-        return jnp.eye(N=shape[0], M=shape[1], dtype=dtype)
-
     @nn.compact
     def __call__(self, x, y):
         """
```
Comment (Algue-Rythme) [inline cnqr/_src/convex.py]: Please document the expected shape of input and of return of the `__call__` method like I did in ` BjorckDense`. Specify the batch in every case for `x` and `y`.

**Review by Algue-Rythme — COMMENTED** [review]
Great work, thank you Yannick !

There is two minor comments remaining regarding a `squeeze` and a documentation. 

Please add imports of `PICNN` in the file [layers.py file](https://github.com/Algue-Rythme/CertifiedQuantileRegression/blob/dev-picnn/cnqr/layers.py).

When it's done squash all your commits into a one named "add PICNN module" with a `git rebase -i` like in [this post](https://stackoverflow.com/a/5189600)

File: cnqr/_src/convex.py (hunk 9 above)
Comment (y-prudent) [reply cnqr/_src/convex.py]: True, I changed the return into `x.reshape(-1,)` and adapted the code in the convex_test.py file accordingly, thanks for the review!

**Review by Algue-Rythme — APPROVED** [review]
LGTM!
